feat(render): dynamic head content

jaymanmdev opened the pull request:
update the JSX render class to not explicitly inject HTML inside of the body tag, allowing for head meta tags and such to be injected and handled by the browser.

jaymanmdev committed:
update the JSX render class to not explicitly inject HTML inside of the body tag, allowing for head meta tags and such to be injected and handled by the browser.

boywithkeyboard reviewed (dismissed):
lgtm

boywithkeyboard commented:
@jaymanmdev seems like the test suite failed because of the changes you made. Please read the [test file](https://github.com/cheetahstudio/cheetah/blob/dev/test/render.test.tsx) and try to fix the issue.

boywithkeyboard commented:
Other than that, I like the idea behind this pr! :)

jaymanmdev committed:
fix last pull request for HTML render changes as it didn't do as I had expected initially.

jaymanmdev committed:
add new render test to check that empty style tag is not injected, slightly change render body assignment to not inject empty style tag if twind isn't utilised.

jaymanmdev commented:
Sorry, this is my first ever PR in any repo so I am still getting the hang of it, I'm getting in the habit of running the tests before submitting an adjustment to my PR/s from now on. Thanks!

Edit: I can merge the two render tests together later on if that would be preferred. :)

boywithkeyboard commented:
![image](https://github.com/cheetahstudio/cheetah/assets/62482066/4cd99860-78bc-44a9-baa3-afb8b1a34da0)

Our [contributing guidelines](https://github.com/azurystudio/cheetah/blob/dev/contributing.md) are really basic but that's the one thing that's included in them. 😅

jaymanmdev commented:
> ![image](https://user-images.githubusercontent.com/62482066/265298344-4cd99860-78bc-44a9-baa3-afb8b1a34da0.png)
> 
> Our [contributing guidelines](https://github.com/azurystudio/cheetah/blob/dev/contributing.md) are really basic but that's the one thing that's included in them. 😅

Ah, I didn't pick up on that guidelines file - my bad! I tested the suite this time though.

boywithkeyboard commented:
Just make a sub-test, e.g.

```ts
Deno.test('something', async (t) => {
  await t.step('first step', async () => {
    // some code
  })
})
```

jaymanmdev commented:
> Just make a sub-test, e.g.
> 
> ```ts
> Deno.test('something', async (t) => {
>   await t.step('first step', async () => {
>     // some code
>   })
> })
> ```

I'll quickly do this now, and submit an adjustment.

jaymanmdev committed:
update render tests to include 3 steps (1 with Twind styling, 1 without Twind styling to test for empty style tag injection, and 1 to test for meta tag injection in head.)

jaymanmdev committed:
move Cheetah app creation to the outside of the test to clean things up a bit.

kibibites commented on render.ts:
```diff
@@ -11,13 +11,30 @@ import renderToString from 'https://esm.sh/preact-render-to-string@6.1.0?deps=pr
     const { html, css } = extract(htmlString)
-
-    c.res.body = `<style>${css}</style><body>${html}</body>`
+    console.log(css.length)
```
remove this please, or only print this when `c.dev` is true

jaymanmdev replied:
oh, I didn't even know that was there still... sorry about that - I was wondering, why I was getting the length in the test logs.

kibibites replied:
all good 😄

kibibites commented on render.ts:
```diff
@@ -11,13 +11,30 @@ import renderToString from 'https://esm.sh/preact-render-to-string@6.1.0?deps=pr
+    console.log(css.length)
+    const startIdxOfHeadContent = html.indexOf(HEAD_TAG)
+    const endIdxOfHeadContent = html.indexOf(
+      `${HEAD_TAG.at(0)}/${HEAD_TAG.slice(1)}`,
```
why not just use `</head>`? :p

jaymanmdev replied:
that's true, I go a bit over the top sometimes lol...

jaymanmdev replied:
Pushed changes for this.

kibibites commented on render.ts:
```diff
@@ -11,13 +11,30 @@ import renderToString from 'https://esm.sh/preact-render-to-string@6.1.0?deps=pr
+    const endIdxOfHeadContent = html.indexOf(
+      `${HEAD_TAG.at(0)}/${HEAD_TAG.slice(1)}`,
+    )
+    const headContent =
```
maybe using regex would be more efficient?

```suggestion
    const headRegex = /<head>([\s\S]*?)<\/head>/g;

    const headContent = headRegex.exec(html)[1];
```

jaymanmdev replied:
regex is generally slower, no?

I don't really use regex personally so I'm not sure.

kibibites replied:
![image](https://github.com/cheetahstudio/cheetah/assets/47074495/62b86b60-b08d-4a3a-b2a8-8fd110e1c3e1)

LOL i never realized regex is that slow

this is all good from me then 😄

jaymanmdev replied:
59 ops/s??? oh my! I heard it was slow... but didn't think it would be THAT slow. lol

jaymanmdev committed:
remove unnecessary console.log in render function, use </head> instead of slicing the HEAD_TAG constant. (changes requested by not-ivy)

jaymanmdev committed:
optimize the 'no styles' case in the render function (check for empty css from Twind straight away, and return the html as is without any changes if that is the case)

jaymanmdev committed:
feat(render): add new render test case for JSX without the html, head or body tags to ensure that the browser and document is still generated appropriately, remove the head manipulation in the render function as the browser will usually know where to put specific tags upon request.

jaymanmdev commented:
b1341059dd16ef47954ad1c94e865c41c836a73e

Let me know what you think.

1. I removed the head manipulation that I was doing as the browser knows where to put specific tags, unless specifically stated otherwise (e.g. putting a meta tag inside the body tag.). We can build out better sanitization of the HTML later on if you'd like?
2. I added a new test case to test this.

boywithkeyboard reviewed (approved):
lgtm
